### ISSUE TITLE
kube-proxy evicted due to low compute resources and not resheduled

### ISSUE THREAD

**Issue by ekuefler** [kubernetes/kubernetes#36539]
**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.1", GitCommit:"33cf7b9acbb2cb7c9c72a10d6636321fb180b159", GitTreeState:"clean", BuildDate:"2016-10-10T18:19:49Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.5", GitCommit:"5a0a696437ad35c133c0c8493f7e9d22b0f9b81b", GitTreeState:"clean", BuildDate:"2016-10-29T01:32:42Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: Google Container Engine
- **OS** (e.g. from /etc/os-release):
```
BUILD_ID=8820.0.0
NAME="Container-VM Image"
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=55
BUG_REPORT_URL=https://crbug.com/new
PRETTY_NAME="Google Container-VM Image"
VERSION=55
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/compute/docs/containers/vm-image/"
ID=gci
```
- **Kernel** (e.g. `uname -a`):
```
Linux gke-prod-highmem-14-884676d5-kx3d 4.4.14+ #1 SMP Tue Sep 20 10:32:07 PDT 2016 x86_64 Intel(R) Xeon(R) CPU @ 2.30GHz GenuineIntel GNU/Linux
```
- **Install tools**: GCP web console

**What happened**:

At some point, the kube-proxy pod on one of my nodes was evicted due to low compute resources and never resheduled on that node. As services were updated, the node was not informed of the new service endpoints, and it began dropping traffic that was routed to it. This caused a percentage of user requests to fail with a "No route to host" error until the node was cordoned.

**What you expected to happen**:

Ideally the kube-proxy pod should be protected from eviction as a core part of the system. If kube-proxy does get evicted, it should get rescheduled when resources are available on the node.

**How to reproduce it** (as minimally and precisely as possible):

Run a cluster with a number of nodes on GKE running at close to capacity until a kube-proxy pod gets evicted.

**Anything else do we need to know**:

It looks to me like kube-proxy is running as a bare pod on each node with nothing to protect it from termination. I'm not sure if it's GKE that set things up this way or if this is how Kubernetes is always configured - if the former this might be a GKE bug. I expected this to run via a DaemonSet to ensure that there is always exactly one instance per node.

I'm also not really sure why the container was evicted in the first place. The node has 100GB of RAM and is running at 97% capacity, which should leave at least a little left over for system processes.

The problem can be observed from kubectl via the following commands:
```
# kubectl get po --namespace=kube-system -a
NAME                                                      READY     STATUS    RESTARTS   AGE
fluentd-cloud-logging-gke-prod-highmem-14-884676d5-kx3d   0/1       Evicted   0          16d
fluentd-cloud-logging-gke-prod-highmem-14-884676d5-n7sh   1/1       Running   1          33d
fluentd-cloud-logging-gke-prod-highmem-14-884676d5-rvm0   0/1       Evicted   0          9d
fluentd-cloud-logging-gke-prod-highmem-14-884676d5-x67e   1/1       Running   0          9d
fluentd-cloud-logging-gke-prod-pri-e1dc3bef-bmb8          1/1       Running   0          35d
fluentd-cloud-logging-gke-prod-pri-e1dc3bef-bxfm          1/1       Running   0          35d
fluentd-cloud-logging-gke-prod-pri-e1dc3bef-mzxf          1/1       Running   0          35d
fluentd-cloud-logging-gke-prod-pri-e1dc3bef-x3my          1/1       Running   1          33d
heapster-v1.2.0-1041656245-gmfz5                          2/2       Running   0          35d
kube-dns-v20-bt6md                                        3/3       Running   0          19d
kube-dns-v20-ttj0q                                        3/3       Running   0          19d
kube-dns-v20-ymw0j                                        3/3       Running   0          19d
kube-proxy-gke-prod-highmem-14-884676d5-kx3d              0/1       Evicted   0          16d
kube-proxy-gke-prod-highmem-14-884676d5-n7sh              1/1       Running   1          33d
kube-proxy-gke-prod-highmem-14-884676d5-rvm0              0/1       Evicted   0          9d
kube-proxy-gke-prod-highmem-14-884676d5-x67e              1/1       Running   0          9d
kube-proxy-gke-prod-pri-e1dc3bef-bmb8                     1/1       Running   0          35d
kube-proxy-gke-prod-pri-e1dc3bef-bxfm                     1/1       Running   0          35d
kube-proxy-gke-prod-pri-e1dc3bef-mzxf                     1/1       Running   0          35d
kube-proxy-gke-prod-pri-e1dc3bef-x3my                     1/1       Running   0          33d
kubedash-2697969532-e51a0                                 2/2       Running   0          33d
kubernetes-dashboard-v1.4.0-sljg9                         1/1       Running   1          35d
l7-default-backend-v1.0-ip2aw                             1/1       Running   0          35d
# kubectl describe po kube-proxy-gke-prod-highmem-14-884676d5-rvm0 --namespace=kube-system
Name:	kube-proxy-gke-prod-highmem-14-884676d5-rvm0
Namespace:	kube-system
Node:	gke-prod-highmem-14-884676d5-rvm0/
Start Time:	Mon, 31 Oct 2016 10:33:34 -0700
Labels:	component=kube-proxy
tier=node
Status:	Failed
Reason:	Evicted
Message:	The node was low on compute resources.
IP:	
Controllers:	<none>
Containers:
  kube-proxy:
    Image:	gcr.io/google_containers/kube-proxy:dce5681a0d470b2efa3e0542b1164c9b
    Port:	
    Command:
      /bin/sh
      -c
      kube-proxy --master=[REDACTED] --kubeconfig=/var/lib/kube-proxy/kubeconfig --cluster-cidr=10.92.0.0/14 --resource-container="" --v=2 1>>/var/log/kube-proxy.log 2>&1
    Requests:
      cpu:	100m
    Volume Mounts:
      /etc/ssl/certs from ssl-certs-host (ro)
      /var/lib/kube-proxy/kubeconfig from kubeconfig (rw)
      /var/log from varlog (rw)
    Environment Variables:	<none>
Volumes:
  ssl-certs-host:
    Type:	HostPath (bare host directory volume)
    Path:	/usr/share/ca-certificates
  kubeconfig:
    Type:	HostPath (bare host directory volume)
    Path:	/var/lib/kube-proxy/kubeconfig
  varlog:
    Type:	HostPath (bare host directory volume)
    Path:	/var/log
QoS Class:	Burstable
Tolerations:	<none>
No events.
```

**Comment by yujuhong** [comment]
Thanks for reporting the issue!


> It looks to me like kube-proxy is running as a bare pod on each node with nothing to protect it from termination. I'm not sure if it's GKE that set things up this way or if this is how Kubernetes is always configured - if the former this might be a GKE bug. I expected this to run via a DaemonSet to ensure that there is always exactly one instance per node.

This is how kuberenetes is configured. Eventually we'd like to move kube-proxy to a DaemonSet, but this is not done yet (#23225).

As for evicting the kube-proxy pod permanently, that's a flaw/bug in the current design, and should be fixed (tracked by #35521).

I am closing the issue since the two problems are being discussed in separate issues already.

**Comment by vgarcia-te** [comment]
I think this can be fixed by marking the Pod as critical: https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/#marking-pod-as-critical
